src/modules/test/main.c

```diff
@@ -0,0 +1,732 @@
+    }
+
+    if (NULL == (test->component = strdup(json_object_get_string(object, RECIPE_COMPONENT))))
+    {
+        LOG_ERROR("'%s' is required for '%s' test step", RECIPE_COMPONENT, type);
+        status = EINVAL;
+    }
+
+    if (NULL == (test->object = strdup(json_object_get_string(object, RECIPE_OBJECT))))
+    {
+        LOG_ERROR("'%s' is required for '%s' test", RECIPE_OBJECT, type);
+        status = EINVAL;
+    }
+
+    if (NULL == (payload = json_object_get_value(object, RECIPE_PAYLOAD)))
+    {
+        if (NULL != (json = json_object_get_string(object, RECIPE_JSON)))
+        {
+            payload = json_parse_string(json);
+        }
+    }
+
+    test->payload = payload ? json_value_deep_copy(payload) : NULL;
+    test->status = json_object_get_number(object, RECIPE_STATUS);
+
+    return status;
+}
+
+int ParseCommandStep(const JSON_Object* object, COMMAND_STEP* command)
+{
+    int status = 0;
+    JSON_Value* value = NULL;
+    const char* arguments = NULL;
+
+    if (command == NULL)
+    {
+        LOG_ERROR("Invalid (null) command");
+        return EINVAL;
+    }
+
+    if (NULL == (value = json_object_get_value(object, RECIPE_RUN_COMMAND)))
+    {
+        LOG_ERROR("Missing '%s' from command step", RECIPE_RUN_COMMAND);
+        status = EINVAL;
+    }
+    else if (JSONString != json_value_get_type(value))
+    {
+        LOG_ERROR("'%s' must be a string", RECIPE_RUN_COMMAND);
+        status = EINVAL;
+    }
+    else if (NULL == (arguments = json_value_get_string(value)))
+    {
+        LOG_ERROR("Failed to serialize '%s' from command step", RECIPE_RUN_COMMAND);
+        status = EINVAL;
+    }
+    else if (NULL == (command->arguments = strdup(arguments)))
+    {
+        LOG_ERROR("Failed to copy '%s' from command step", RECIPE_RUN_COMMAND);
+        status = ENOMEM;
+    }
+
+    command->status = json_object_get_number(object, "Status");
+
+    return status;
+}
+
+int ParseModuleStep(const JSON_Object* object, MODULE_STEP* module)
+{
+    int status = 0;
+    const char* action = NULL;
+
+    if (module == NULL)
+    {
+        LOG_ERROR("Invalid (null) module");
+        return EINVAL;
+    }
+
+    if (NULL == (action = json_object_get_string(object, RECIPE_ACTION)))
+    {
+        LOG_ERROR("Missing '%s' from module step", RECIPE_ACTION);
+        status = EINVAL;
+    }
+    else
+    {
+        if (strcmp(action, RECIPE_LOAD_MODULE) == 0)
+        {
+            module->action = LOAD;
+        }
+        else if (strcmp(action, RECIPE_UNLOAD_MODULE) == 0)
+        {
+            module->action = UNLOAD;
+        }
+        else
+        {
+            LOG_ERROR("Invalid action '%s'", action);
+            status = EINVAL;
+        }
+    }
+
+    if ((status == 0) && (module->action == LOAD))
+    {
+        if (NULL == (module->name = strdup(json_object_get_string(object, RECIPE_MODULE))))
+        {
+            LOG_ERROR("'%s' is required for '%s' module step", RECIPE_MODULE, action);
```

Comment:
>module step [](http://example.com/codeflow?start=49&length=11)

What's a "module step"? May want to use in logging only terms that are defined in modules.md.